### PR TITLE
Fix ErgoStateContextSpec Sporadic Failures

### LINKED ISSUE #2114: ErgoStateContextSpec is unstable
`ErgoStateContextSpec` fails very rarely & sporadically.

this came up during work on

- https://github.com/ergoplatform/ergo/pull/2109

### Code

https://github.com/ergoplatform/ergo/blob/master/src/test/scala/org/ergoplatform/nodeView/state/ErgoStateContextSpec.scala

### Tasks

- [x] create a reproducer
  - https://github.com/ergoplatform/ergo/pull/2116
- [x] isolate bug
  - pinpointed to `ErgoGenerators.extensionKvGen()`
- [x] fix test (fix/workaround) bugs
  - https://github.com/ergoplatform/ergo/pull/2116

### Followup Tasks

- https://github.com/ergoplatform/ergo/issues/2117
- https://github.com/ergoplatform/ergo/issues/2118

### Similar Issues

- #2095

### PR DESCRIPTION
**Pull request by stenolog**
closes #2114

### CODE REVIEW

**Comment by stenolog** [comment]
@kushti , this should be ready to go in.

I can squash from my side, or you could enable UI squashing

https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/configuring-commit-squashing-for-pull-requests

and do it from your side.

**Comment by kushti** [comment]
Good catch!

For 

```
var imvValue = extensionKvGen(Extension.FieldKeySize, Extension.FieldValueMaxSize + 1).sample.get
    while (imvValue._1.head == 0) {
        imvValue = extensionKvGen(Extension.FieldKeySize, Extension.FieldValueMaxSize + 1).sample.get
    }
    sc.appendFullBlock(fbWithFields(imvValue +: oldFields)) shouldBe 'failure
```

the issue is that `extensionKvGen` is defined as 

```
  def extensionKvGen(keySize: Int, valuesSize: Int): Gen[(Array[Byte], Array[Byte])] = for {
    key <- genSecureBoundedBytes(keySize, keySize)
    value <- if (key.head == 0) genSecureBoundedBytes(4, 4) else genSecureBoundedBytes(valuesSize, valuesSize)
  } yield (key, value)
```

so when `imvValue._1.head == 0`, value is 4 bytes long, not 65 (Extension.FieldValueMaxSize + 1)

I think dedicated generator should be used for the check

**Comment by kushti** [comment]
for 
```
var validMKV = extensionKvGen(Extension.FieldKeySize, Extension.FieldValueMaxSize).sample.get
while (validMKV._1.head != 1) {
  validMKV = extensionKvGen(Extension.FieldKeySize, Extension.FieldValueMaxSize).sample.get
}
```

generation of test data here is good, as for `validMKV._1.head == 1` interlinks would be packed improperly, and interlinks packing/unpacking is tested in another test

**Comment by stenolog** [comment]
> while (validMKV._1.head != 1) {

It's 'validMKV._1.head == 1'

This fails, too, there are two different sporadic failures.

**Comment by stenolog** [comment]
> I think dedicated generator should be used for the check

My thought, too, see https://github.com/ergoplatform/ergo/issues/2118 

"Possibly introduce extensionKvGenValidated()"

This PR/iteration can/should go in as-is, fixes the 2 different sporadic test-failures.

Next step/iteration/PR would be to clarify/rewrite/document

- https://github.com/ergoplatform/ergo/issues/2117
- https://github.com/ergoplatform/ergo/issues/2118

**Comment by kushti** [comment]
@stenolog remove 

```
printf("\r%d05 ", i)
```
and `for` loop added, not related to fixes

**Comment by stenolog** [comment]
> remove
> 
> ```
> printf("\r%d05 ", i)
> ```

this can be remove, yes.

> and `for` loop added, not related to fixes

This is related because the test needs to loop through the randomness to catch any future regressions.

Without it, new similar bugs can slip through undetected.

**Comment by stenolog** [comment]
> and `for` loop added, not related to fixes

I've removed (out-commented) it.

File: src/test/scala/org/ergoplatform/nodeView/state/ErgoStateContextSpec.scala [hunk 1]
```diff
@@ -58,13 +58,27 @@ class ErgoStateContextSpec extends HistoryTestHelpers {
       fb.copy(extension = newExtension)
     }
 
+    // regression stress-test for sporadic failure https://github.com/ergoplatform/ergo/issues/2114
+    // disabling for now
+    // for (i <- 1 to 1000) {
```
Comment (kushti) [inline src/test/scala/org/ergoplatform/nodeView/state/ErgoStateContextSpec.scala:63]: please avoid commenting out code

File: src/test/scala/org/ergoplatform/nodeView/state/ErgoStateContextSpec.scala [hunk 2]
```diff
@@ -73,13 +87,23 @@ class ErgoStateContextSpec extends HistoryTestHelpers {
     ).fields
     sc.appendFullBlock(fbWithFields(invalidInterlinks ++ oldFields)) shouldBe 'failure
 
+    // https://github.com/ergoplatform/ergo/issues/2114
+    // if validMKV._1.head is 1, appendFullBlock within "valid application of correct extension" will fail
+    // As a workaround, just generate new values until (validMKV._1.head != 1)
+    // TODO: investigate and provide a full fix (followup issue)
```
Comment (kushti) [inline src/test/scala/org/ergoplatform/nodeView/state/ErgoStateContextSpec.scala:93]: this fix is okay here, just add a note that with key.head == 1 improperly packed interlink would be generated

File: src/test/scala/org/ergoplatform/nodeView/state/ErgoStateContextSpec.scala [hunk 3]
```diff
@@ -58,13 +58,27 @@ class ErgoStateContextSpec extends HistoryTestHelpers {
...
+
     // checks, specific for extension
     // validation of field keys size
     val imvKey = extensionKvGen(Extension.FieldKeySize - 1, Extension.FieldValueMaxSize).sample.get
     sc.appendFullBlock(fbWithFields(imvKey +: oldFields)) shouldBe 'failure
 
+    // https://github.com/ergoplatform/ergo/issues/2114
+    // this fails sporadically, when `(imvValue._1.head == 0)`, because less value bytes will be generated
+    // by extensionKvGen(). Workaround is to just generate again while `(imvValue._1.head == 0)`
+    // TODO: document, and possibly rewrite extensionKvGen after the above issues are clarified/solved
```
Comment (kushti) [inline src/test/scala/org/ergoplatform/nodeView/state/ErgoStateContextSpec.scala:74]: likely 4 bytes output was made for a reason, so better to use another Array[Byte] generator here

File: src/test/scala/org/ergoplatform/nodeView/state/ErgoStateContextSpec.scala [hunk 4]
```diff
@@ -73,13 +87,23 @@ class ErgoStateContextSpec extends HistoryTestHelpers {
...
-    val validMKV = extensionKvGen(Extension.FieldKeySize, Extension.FieldValueMaxSize).sample.get
     sc.appendFullBlock(fbWithFields(Seq(validMKV, validMKV) ++ oldFields)) shouldBe 'failure
 
     // valid application of correct extension
     sc.appendFullBlock(fbWithFields(validMKV +: oldFields)) shouldBe 'success
 
-  }
+    }
+
+  //} // test-loop
```
Comment (kushti) [inline src/test/scala/org/ergoplatform/nodeView/state/ErgoStateContextSpec.scala:107]: commented out code

**Review by kushti — CHANGES_REQUESTED** [review]

File: src/test/scala/org/ergoplatform/nodeView/state/ErgoStateContextSpec.scala (hunk 3 above)
Comment (stenolog) [reply src/test/scala/org/ergoplatform/nodeView/state/ErgoStateContextSpec.scala:74]: > likely 4 bytes output was made for a reason

...which is undocumented.

> better to use another Array[Byte] generator here

already mentioned within:

- https://github.com/ergoplatform/ergo/issues/2118

"Possibly introduce extensionKvGenValidated()"

(work on https://github.com/ergoplatform/ergo/issues/2118 should show what is the prefered way to go, documentation needs to be added there anyways)

File: src/test/scala/org/ergoplatform/nodeView/state/ErgoStateContextSpec.scala (hunk 1 above)
Comment (stenolog) [reply src/test/scala/org/ergoplatform/nodeView/state/ErgoStateContextSpec.scala:63]: removed

File: src/test/scala/org/ergoplatform/nodeView/state/ErgoStateContextSpec.scala (hunk 4 above)
Comment (stenolog) [reply src/test/scala/org/ergoplatform/nodeView/state/ErgoStateContextSpec.scala:107]: removed

File: src/test/scala/org/ergoplatform/nodeView/state/ErgoStateContextSpec.scala (hunk 2 above)
Comment (stenolog) [reply src/test/scala/org/ergoplatform/nodeView/state/ErgoStateContextSpec.scala:91]: added

File: src/test/scala/org/ergoplatform/nodeView/state/ErgoStateContextSpec.scala (hunk 3 above)
Comment (stenolog) [reply src/test/scala/org/ergoplatform/nodeView/state/ErgoStateContextSpec.scala]: changed "possibly rewrite" to "possibly rewrite/replace"

**Review by kushti — CHANGES_REQUESTED** [review]
I think custom generators should be used , reiterating over extensionKvGen is messy . As your goal is to avoid specific key spaces anyway, better to do that directly

**Comment by stenolog** [comment]
> As your goal is

My goal was to detect the sporadic bug (done) and fix it (done).

Further work was planned within

* https://github.com/ergoplatform/ergo/issues/2117
* https://github.com/ergoplatform/ergo/issues/2118

which you've closed prematurely.

It is very(!) difficult to produce quality results here (waiting for nearly a week, then change-request. Waiting 5 days, change-request).

Writing this just to file my objections.

I'll provide a "quick&dirty" generator, not because I agree, but because I'm forced to.

**Comment by stenolog** [comment]
> I'll provide a "quick&dirty" generator,

I've problems with this (conditionals within generator functions)

### Problem

to avoid this messy construct

```scala
    var validMKV = extensionKvGen(Extension.FieldKeySize, Extension.FieldValueMaxSize).sample.get
    while (validMKV._1.head == 1) {
      validMKV = extensionKvGen(Extension.FieldKeySize, Extension.FieldValueMaxSize).sample.get
    }
```

a dedicated generator should be provided.

Problem is that conditionals within the generator seem not trivial in scala (found no docs/articles)

```sh
  def extensionKvGen(keySize: Int, valuesSize: Int): Gen[(Array[Byte], Array[Byte])] = for {
   # key.head should NOT be 1
    key <- genSecureBoundedBytes(keySize, keySize)
    value <- genSecureBoundedBytes(valuesSize, valuesSize)
  } yield (key, value)
```

Any suggestion on how to check for `key == 1` within the generator, to repeat the generation?

**Comment by stenolog** [comment]
> Any suggestion on how to check for `key == 1` within the generator, to repeat the generation?

from https://discord.com/channels/668903786361651200/669989266478202917/1214238632383217715

```scala
def extensionKvGen(keySize: Int, valueSize: Int): Gen[(Array[Byte], Array[Byte])] = {
  def genSecureBoundedBytes(size: Int): Gen[Array[Byte]] = {
    // Your existing implementation to generate Array[Byte] of given size.
  }

  def keyGen: Gen[Array[Byte]] = genSecureBoundedBytes(keySize).flatMap { key =>
    if (key.headOption.contains(1.toByte)) keyGen else Gen.const(key)
  }

  for {
    key <- keyGen
    value <- genSecureBoundedBytes(valueSize)
  } yield (key, value)
}
```

**Comment by stenolog** [comment]
@kushti , in order to speed things up, please feel free to modify/extend this PR as you see fit.

I'm fine with it, even if `stenolog` is not shown as the committer.

As a sidenote: this (sporadic test bug) PR remained completely unprocessed, despite several pings:

- https://github.com/ergoplatform/ergo/pull/2121

**Review by kushti — APPROVED** [review]

**Pull request merged**